### PR TITLE
Feature/get post toggle bookmark

### PR DESCRIPTION
**Pull request by liruchen32**
# Issue
[[維運] redux 升級後續 refactor: 書籤](https://app.asana.com/0/1203699264568808/1207639742964814/f)

# Notice

# Dependency

### CODE REVIEW

File: controllers/news_v2.go [hunk 1]
```diff
@@ -123,6 +124,25 @@ func (nc *newsV2Controller) GetAPost(c *gin.Context) {
 		if len(posts) > 0 {
 			post = posts[0]
 		}
+
+		if fullPost, ok := post.(news.Post); ok {
```
Comment (Aylie-Chou) [inline controllers/news_v2.go:128]: Question: why do we need to add type convertion (news.Post) here ?

File: internal/news/query.go [hunk 2]
```diff
@@ -243,6 +245,10 @@ func parseSingleQuery(c *gin.Context) *Query {
 	if full, err := strconv.ParseBool(c.Query(queryFull)); err == nil {
 		q.Full = full
 	}
+
+	if toggleBookmark, err := strconv.ParseBool(c.Query(queryToggleBookmark)); err == nil {
+		q.ToggleBookmark = toggleBookmark
```
Comment (Aylie-Chou) [inline internal/news/query.go:250]: Could we also add `q.ToggleBookmark` in `ParsePostListQuery` ?

File: storage/news_v2.go [hunk 3]
```diff
@@ -95,6 +95,54 @@ func (gs *gormStorage) GetBookmarksOfPostsTask(ctx context.Context, userID strin
 	return result
 }
 
+func (gs *gormStorage) GetBookmarksForFullPost(ctx context.Context, userID string, post news.Post) (models.UsersBookmarks, error) {
+	if userID == "" {
+		log.Println("userID is required")
+		return models.UsersBookmarks{}, errors.New("userID is required")
```
Comment (Aylie-Chou) [inline storage/news_v2.go:101]: Question: why do we choose to return empty struct rather than `nil` when error ?

File: storage/news_v2.go [hunk 4]
```diff
@@ -95,6 +95,54 @@ func (gs *gormStorage) GetBookmarksOfPostsTask(ctx context.Context, userID strin
...
+	case result, ok := <-gs.GetBookmarksForFullPostTask(ctx, userID, post):
+		if !ok {
+			return models.UsersBookmarks{}, errors.New("failed to get bookmark")
+		}
+		if result.Error != nil {
+			return models.UsersBookmarks{}, result.Error
+		}
+
+		log.Print(result)
+		targetBookmark = result.Content.(models.UsersBookmarks)
```
Comment (Aylie-Chou) [inline storage/news_v2.go:118]: targetBookmark, ok ?

File: storage/news_v2.go [hunk 5]
```diff
@@ -95,6 +95,54 @@ func (gs *gormStorage) GetBookmarksOfPostsTask(ctx context.Context, userID strin
...
+		return models.UsersBookmarks{}, ctx.Err()
+	case result, ok := <-gs.GetBookmarksForFullPostTask(ctx, userID, post):
+		if !ok {
+			return models.UsersBookmarks{}, errors.New("failed to get bookmark")
+		}
+		if result.Error != nil {
+			return models.UsersBookmarks{}, result.Error
+		}
+
+		log.Print(result)
```
Comment (Aylie-Chou) [inline storage/news_v2.go:117]: remove debug log XD

File: storage/news_v2.go [hunk 6]
```diff
@@ -95,6 +95,54 @@ func (gs *gormStorage) GetBookmarksOfPostsTask(ctx context.Context, userID strin
 	return result
 }
 
+func (gs *gormStorage) GetBookmarksForFullPost(ctx context.Context, userID string, post news.Post) (models.UsersBookmarks, error) {
+	if userID == "" {
```
Comment (Aylie-Chou) [inline storage/news_v2.go:99]: Question: do we also need to check post.ID exists ?

File: storage/news_v2.go [hunk 7]
```diff
@@ -95,6 +95,54 @@ func (gs *gormStorage) GetBookmarksOfPostsTask(ctx context.Context, userID strin
...
+	result := make(chan fetchResult)
+	go func(ctx context.Context, userID string, post news.Post) {
+
+		postID := post.ID.Hex()
+
+		var UserBookmark models.UsersBookmarks
+		err := gs.db.Where("user_id = ? AND post_id = ?", userID, postID).Find(&UserBookmark).Error
+
+		if err != nil {
+			log.WithField("detail", err).Errorf("%s", f.FormatStack(err))
```
Comment (Aylie-Chou) [inline storage/news_v2.go:138]: error could be returned by `fetchResult{Error: errors.WithStack(err)}` & check `result.Error != nil` in receiver (looks like the receiver has already catch it)

File: storage/news_v2.go [hunk 8]
```diff
@@ -95,6 +95,54 @@ func (gs *gormStorage) GetBookmarksOfPostsTask(ctx context.Context, userID strin
...
+
+
+func (gs *gormStorage) GetBookmarksForFullPostTask(ctx context.Context, userID string, post news.Post) <-chan fetchResult {
+	result := make(chan fetchResult)
+	go func(ctx context.Context, userID string, post news.Post) {
+
+		postID := post.ID.Hex()
+
+		var UserBookmark models.UsersBookmarks
+		err := gs.db.Where("user_id = ? AND post_id = ?", userID, postID).Find(&UserBookmark).Error
```
Comment (Aylie-Chou) [inline storage/news_v2.go:135]: (nice to have) Could we also match `post_id` rather than `slug` in `GetBookmarksOfPosts` XD

**Review by Aylie-Chou — CHANGES_REQUESTED** [review]
as comments~~

File: controllers/news_v2.go (hunk 1 above)
Comment (liruchen32) [reply controllers/news_v2.go:128]: cause `post` type is interface{}

File: internal/news/query.go (hunk 2 above)
Comment (liruchen32) [reply internal/news/query.go:250]: OK

File: storage/news_v2.go (hunk 6 above)
Comment (liruchen32) [reply storage/news_v2.go:99]: already check

File: storage/news_v2.go (hunk 3 above)
Comment (liruchen32) [reply storage/news_v2.go:101]: cannot use nil as models.UsersBookmarks value in return statement

File: storage/news_v2.go (hunk 5 above)
Comment (liruchen32) [reply storage/news_v2.go:117]: fixed

File: storage/news_v2.go (hunk 4 above)
Comment (liruchen32) [reply storage/news_v2.go:118]: fixed

File: storage/news_v2.go (hunk 7 above)
Comment (liruchen32) [reply storage/news_v2.go:138]: fixed

File: storage/news_v2.go (hunk 3 above)
Comment (Aylie-Chou) [reply storage/news_v2.go:101]: okay, I think we could refactor it to return the pointer version in the future.

**Review by Aylie-Chou — APPROVED** [review]
LGTM

**Pull request merged**